### PR TITLE
[improve][broker] System topic writer/reader connection not counted.

### PR DESCRIPTION
**Pull request by Technoboy-**
Master Issue: #18364

### Motivation

It's better not to count the system topic writer/reader connection.

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

**Comment by grssam** [comment]
@Technoboy-  Thank you for the PR.
It's weird that such a generic issue, and we never faced it in our production system till now.. Is this a regression in 2.9.3? We were previously on 2.7.4 up until september. I tried to look at differences in 2.7 vs 2.9 wrt this particular context and could not see any sort of regression; Does it mean that the problem existed in 2.7.* as well?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java [hunk 1]
```diff
@@ -439,6 +439,9 @@ private PublishRate publishRateInBroker(ServiceConfiguration config) {
     }
 
     protected boolean isProducersExceeded() {
+        if (isSystemTopic()) {
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java:442]: Hmm, shall we need to skip the replicator also?

**Comment by Technoboy-** [comment]
> @Technoboy- Thank you for the PR. It's weird that such a generic issue, and we never faced it in our production system till now.. Is this a regression in 2.9.3? We were previously on 2.7.4 up until september. I tried to look at differences in 2.7 vs 2.9 wrt this particular context and could not see any sort of regression; Does it mean that the problem existed in 2.7.* as well?

yes, it existed in branch-2.7.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/systopic/PartitionedSystemTopicTest.java [hunk 2]
```diff
@@ -260,4 +265,61 @@ private void testSetBacklogCausedCreatingProducerFailure() throws Exception {
...
+        NamespaceEventsSystemTopicFactory systemTopicFactory = new NamespaceEventsSystemTopicFactory(pulsarClient);
+        TopicPoliciesSystemTopicClient systemTopicClientForNamespace = systemTopicFactory
+                .createTopicPoliciesSystemTopicClient(NamespaceName.get(ns));
+        SystemTopicClient.Reader reader1 = systemTopicClientForNamespace.newReader();
+        SystemTopicClient.Reader reader2 = systemTopicClientForNamespace.newReader();
+
+        admin.topicPolicies().setMaxProducers(topic, 1);
+        AtomicInteger successCount = new AtomicInteger(0);
+        CountDownLatch latch = new CountDownLatch(3);
+        new Thread(() -> {
```
[the remark this answers is not in the document]
Comment (Technoboy-) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/systopic/PartitionedSystemTopicTest.java]: Ah, fixed. update test.

**Comment by grssam** [comment]
> > @Technoboy- Thank you for the PR. It's weird that such a generic issue, and we never faced it in our production system till now.. Is this a regression in 2.9.3? We were previously on 2.7.4 up until september. I tried to look at differences in 2.7 vs 2.9 wrt this particular context and could not see any sort of regression; Does it mean that the problem existed in 2.7.* as well?
> 
> yes, it existed in branch-2.7.

Thanks. It's weird that no one ever faced this issue before.. even we have been running for over an year without facing this issue..

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java (hunk 1 above)
Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java]: updated.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorTest.java [hunk 3]
```diff
@@ -1505,4 +1505,34 @@ public void testWhenUpdateReplicationCluster() throws Exception {
...
+        MessageProducer producer1 = new MessageProducer(url1, dest1);
+        log.info("--- Starting producer1 --- " + url1);
+
+        producer1.produce(1);
+
+        @Cleanup
+        MessageProducer producer2 = new MessageProducer(url2, dest2);
+        log.info("--- Starting producer2 --- " + url2);
+
+        producer2.produce(1);
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorTest.java:1536]: It's better to try to add one more producer to make sure new producer is not allowed to connect to the topic.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java [hunk 4]
```diff
@@ -438,14 +438,21 @@ private PublishRate publishRateInBroker(ServiceConfiguration config) {
...
-        if (maxProducers > 0 && maxProducers <= producers.size()) {
+        if (maxProducers != null && maxProducers > 0 && maxProducers <= getUserCreatedProducerSize()) {
             return true;
         }
         return false;
     }
 
+    private long getUserCreatedProducerSize() {
+        return producers.values().stream().filter(p -> !(p.isRemote() || p.getTopic().isSystemTopic())).count();
+    }
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java:454]: Why do we need this change? All the producers under a topic instance should only point to this topic, right?

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/systopic/PartitionedSystemTopicTest.java [hunk 5]
```diff
@@ -260,4 +261,33 @@ private void testSetBacklogCausedCreatingProducerFailure() throws Exception {
             Assert.fail("failed to create producer");
         }
     }
+
+    @Test
+    private void testSystemTopicNotCheckExceed() throws Exception {
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/systopic/PartitionedSystemTopicTest.java:266]: We should also test the max producer limitation for the system topic.

**Review by codelipenghui — COMMENTED** [review]
Do we need to change method `isConsumersExceededOnTopic`? We have max subscriptions and max consumers limitation for a topic.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java (hunk 4 above)
Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java:454]: All the producers (including user-created producers, the system topic producers, and replicator producers) stored in `producers `.  So when checking the user-created producer size, we should filter system topic producers and replicator producers.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorTest.java (hunk 3 above)
Comment (Technoboy-) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorTest.java:1536]: ok

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/systopic/PartitionedSystemTopicTest.java (hunk 5 above)
Comment (Technoboy-) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/systopic/PartitionedSystemTopicTest.java:266]: Ah, I used to set policies to mock the producer. I should create writers directly...
Fixed.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java (hunk 4 above)
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java:454]: Ok, but `p.getTopic().isSystemTopic()` is not required? Line 442 is already handled.

Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java:454]: fixed

**Review by codelipenghui — APPROVED** [review]

**Comment by grssam** [comment]
Please cherry-pick this and #18603 to 2.9 branch as well.

**Pull request merged**